firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -364,41 +450,60 @@ private void deleteIndexEntry(Document document, IndexEntry indexEntry) {
-          fieldIndex,
-          target,
-          arrayValues,
-          lowerBound,
-          upperBound);
-    }
-
-    Object[] lowerBoundEncoded = encodeBound(fieldIndex, target, lowerBound);
-    String lowerBoundOp = lowerBound != null && lowerBound.isInclusive() ? ">=" : ">";
-    Object[] upperBoundEncoded = encodeBound(fieldIndex, target, upperBound);
-    String upperBoundOp = upperBound != null && upperBound.isInclusive() ? "<=" : "<";
-    Object[] notInEncoded = encodeValues(fieldIndex, target, notInValues);
-
-    SQLitePersistence.Query query =
-        generateQuery(
-            target,
-            fieldIndex.getIndexId(),
+    List<String> subQueries = new ArrayList<>();
+    List<Object> bindings = new ArrayList<>();
+
+    for (Target subTarget : getSubTargets(target)) {
+      FieldIndex fieldIndex = getFieldIndex(subTarget);
+      @Nullable List<Value> arrayValues = subTarget.getArrayValues(fieldIndex);
+      @Nullable List<Value> notInValues = subTarget.getNotInValues(fieldIndex);
+      @Nullable Bound lowerBound = subTarget.getLowerBound(fieldIndex);
+      @Nullable Bound upperBound = subTarget.getUpperBound(fieldIndex);
+
+      if (Logger.isDebugEnabled()) {
+        Logger.debug(
+            TAG,
+            "Using index '%s' to execute '%s' (Arrays: %s, Lower bound: %s, Upper bound: %s)",
+            fieldIndex,
+            subTarget,
             arrayValues,
-            lowerBoundEncoded,
-            lowerBoundOp,
-            upperBoundEncoded,
-            upperBoundOp,
-            notInEncoded);
+            lowerBound,
+            upperBound);
+      }
+
+      Object[] lowerBoundEncoded = encodeBound(fieldIndex, subTarget, lowerBound);
+      String lowerBoundOp = lowerBound != null && lowerBound.isInclusive() ? ">=" : ">";
+      Object[] upperBoundEncoded = encodeBound(fieldIndex, subTarget, upperBound);
+      String upperBoundOp = upperBound != null && upperBound.isInclusive() ? "<=" : "<";
+      Object[] notInEncoded = encodeValues(fieldIndex, subTarget, notInValues);
+
+      Object[] subQueryAndBindings =
+          generateQueryAndBindings(
+              subTarget,
+              fieldIndex.getIndexId(),
+              arrayValues,
+              lowerBoundEncoded,
+              lowerBoundOp,
+              upperBoundEncoded,
+              upperBoundOp,
+              notInEncoded);
+      subQueries.add(String.valueOf(subQueryAndBindings[0]));
+      bindings.addAll(Arrays.asList(subQueryAndBindings).subList(1, subQueryAndBindings.length));
+    }
+
+    String queryString =
+        subQueries.size() == 1
+            ?
+            // If there's only one subQuery, just execute the one subQuery.
+            subQueries.get(0)
+            :
+            // Construct "subQuery1 UNION subQuery2 UNION ... LIMIT N"
+            TextUtils.join(" UNION ", subQueries) + " LIMIT " + target.getLimit();
```

Comment:
We need to make sure that we do not pass more than 999 bind elements. The best way to do this probably to validate that the query only contains 100 DNF terms (or what we decide the limit is).

Reply:
Can you also add parenthesis to indicate whether the LIMIT belongs to all statements or just the last one?